networkx/generators/social.py

```diff
@@ -11,12 +11,18 @@
 ]
 
 
-def karate_club_graph():
+def karate_club_graph(weighted=False):
```

Comment:
I would prefer to make it weighted  -- and to not make it optional (no keyword).
The routines can use `weighted=None` to handle the unweighted version.
This is a good addition.  It also simplifies the code a little below (when adding edges) 
if we remove the option.

Reply:
I started working on this, but, unfortunately, there are some complications in fixing two of the three broken tests:

A minor issue is that in [algorithms.community.tests.test_modularity_max.test_modularity_communities](https://github.com/networkx/networkx/blob/cceb43d15e1d01476c8c15ff273399dee0e3b1aa/networkx/algorithms/community/tests/test_modularity_max.py#L10), both greedy_modularity_communities and naive_greedy_modularity_communities are tested, but the latter does not take a `weight` argument since it is a naive reference implementation. The simplest fix is to set the edge weights in the graph to be 1 after loading it instead of trying to call one function with a `weight` argument and the other without.

A larger issue came up when trying to fix [algorithms.tests.test_non_randomness](https://github.com/networkx/networkx/blob/cceb43d15e1d01476c8c15ff273399dee0e3b1aa/networkx/algorithms/tests/test_non_randomness.py#L8). The [non_randomness](https://github.com/networkx/networkx/blob/cceb43d15e1d01476c8c15ff273399dee0e3b1aa/networkx/algorithms/non_randomness.py#L11) function doesn't take a weight parameter, but implicitly uses the weights from the adjacency matrix when computing the graph's eigenvalues by calling `nx.to_numpy_array` without setting a value for `weight` . The seemingly simple fix would be to add a `weight` argument to `non_randomness` and pass it along when getting the adjacency matrix. However, referring to [the paper](http://csce.uark.edu/~xintaowu/publ/299_Ying.pdf) where matrix non-randomness is defined, the graph is explicitly defined as unweighted. I think their calculation naively generalizes to weighted graphs, but I'm not an expert in spectral graph theory, so I'm not completely sure that it does. If it doesn't, then the fix is to always pass `weight=None` when loading the adjacency matrix. Either way, I think a proper fix for the broken test case involves fixing this function and not modifying the test case itself, though it should probably be a topic for a separate issue/pull request since it doesn't directly relate to the karate graph.

The third broken test case, `algorithms.community.tests.test_louvain.test_partition` can be fixed by simply setting `weight=None`, as you suggested.

Reply:
Nice summary!  

The minor issue with `naive_greedy_modularity_communities` is a little embarrassing. That should definitely have a `weight` keyword argument.  I believe the history is that it was a private function for a long time and when we converted it to a public function we forgot to think about the general input structure. I think we should add a keyword with default:  `weight=None`.  

For the second `non_randomness` test, the naive extension from unweighted edges to weighted edges generalizes nicely in this case. Here, we should probably implement a default of `weight="weight"` to avoid backward incompatibility.  So the steps will be:  1) adding a keyword argument to `non_randomness` with default `weight="weight"`. 2) change the test to call it with `weight=None`.

I think these are aligned with your thoughts above.  If you'd rather they be done by someone who has experience with those two functions, I'll take a stab at the changes to `non_randomness` and `naive_greedy_modularity_comunities`. Just let me know.

Reply:
No worries, I went ahead and removed the weight parameter from karate_club_graph, added the weight parameters to non_randomness and naive_greedy_modularity_communities, and updated test cases/documentation accordingly. Hopefully, the changes should be easy to verify, and this is now safe to merge.